### ISSUE TITLE
Clean up a long running script before running it again

### ISSUE THREAD

**Issue by mathieudutour** [skpm/skpm#142]
Ref https://github.com/BohemianCoding/SketchAPI/pull/165

When you are working with a webview for example, the workflow becomes less than ideal: you need to restart sketch everytime you change something.

It would be nice if, instead of directly running the plugin, we could clean the previous one before running the new one.

This probably involves some changes in Sketch itself tho

**Comment by robintindale** [comment]
> long-running script has finished running but hasn't been cleaned up yet and you run the same script again: it will re-use the same JSContext which is due to shut down.

Couldn't this be solved by adding a force-poll at the point when the script is run again, checking whether the context should be shut down (as there are no more long-running elements) or re-used? Or is that what you're proposing with *"clean the previous"*?

Would this be best to be an explicit / manual thing - seeing as killing the long-running element of a plugin every time it's run might be not what people want? As mentioned `teardownSession` seems to do the job, with the caveat that the `onCleanup` functions for fibers aren't called.

**Comment by mathieudutour** [comment]
> Couldn't this be solved by adding a force-poll at the point when the script is run again, checking whether the context should be shut down (as there are no more long-running elements) or re-used? Or is that what you're proposing with "clean the previous"?

That could be a solution yes. I still feel that cleaning the plugin when it's actually done running is cleaner than polling, even if we try to be smarter with the poll.

> Would this be best to be an explicit / manual thing - seeing as killing the long-running element of a plugin every time it's run might be not what people want?

Hum good question. My first thought was to integrate that with the `--run` option. I guess if you don't want to kill the plugin, you wouldn't use `--run` and run the plugin yourself. I'm struggling to see use-cases for it tho?

**Comment by robintindale** [comment]
> cleaning the plugin when it's actually done running is cleaner than polling

yep true, but isn't that a separate discussion - even if the plugin is cleaned up efficiently, it doesn't change the fact that the context is re-used if the plugin hasn't finished execution

So I just tested `skpm-build --watch --run` and it seems to get stuck if the plugin is long-running. My test plugin is
```
  var fiber = require('sketch/async').createFiber()
  console.log('hey');
```
and `skpm-build` hangs, no longer rebuilding. Is this part of the problem?

If I `control+c` and try another `--watch --run`, `sketchtool` crashes Sketch with 
```
2018-04-10 17:44:18.655 sketchtool[59634:4761699] *** Terminating app due to uncaught exception 'NSInvalidReceivePortException', reason: 'connection went invalid while waiting for a reply because a mach port died'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff27e0e00b __exceptionPreprocess + 171
	1   libobjc.A.dylib                     0x00007fff4e9eec76 objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff27e9fc9d +[NSException raise:format:] + 205
	3   Foundation                          0x00007fff29ed3f2d -[NSConnection sendInvocation:internal:] + 2134
	4   CoreFoundation                      0x00007fff27d8412c ___forwarding___ + 748
	5   CoreFoundation                      0x00007fff27d83db8 _CF_forwarding_prep_0 + 120
	6   sketchtool                          0x000000010e7f1585 -[BCSTRunCommand waitForLaunchOfApplicationWithEngine:] + 581
	7   Foundation                          0x00007fff29eca3a5 __NSFireDelayedPerform + 409
	8   CoreFoundation                      0x00007fff27d91bb4 __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ + 20
	9   CoreFoundation                      0x00007fff27d91827 __CFRunLoopDoTimer + 1095
	10  CoreFoundation                      0x00007fff27d9132a __CFRunLoopDoTimers + 346
	11  CoreFoundation                      0x00007fff27d8892b __CFRunLoopRun + 2427
	12  CoreFoundation                      0x00007fff27d87d23 CFRunLoopRunSpecific + 483
	13  CoreFoundation                      0x00007fff27dc6773 CFRunLoopRun + 99
	14  sketchtool                          0x000000010e7b824c ECCommandLineMain + 293
	15  sketchtool                          0x000000010e7f4d01 main + 33
	16  libdyld.dylib                       0x00007fff4f5de115 start + 1
	17  ???                                 0x0000000000000005 0x0 + 5
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

**Comment by mathieudutour** [comment]
oh I forgot about that, sketch now wait for a plugin to finish before returning.

Yeah that will need some thinking to find a good solution.

As a side note, I push new version for the different plugins to cleanly handle an early shut down (for the webview for example, it will be closed if the fiber is cleaned up)

**Comment by LeonardPauli** [comment]
I found a workflow for developing long-running scripts that works without glitches for me (tested with timeouts and skpm/...webview, etc):

__Manual reload of long-running scripts__

- disable script caching
  
  ```sh
  defaults write ~/Library/Preferences/com.bohemiancoding.sketch3.plist AlwaysReloadScript -bool YES
  ```

- add tear down hook

    ```js
    const ver = 1
    coscript.setShouldKeepAround(true)
    console.log('STARTED: '+ver)

    // do long-running things

    const tearDown = ()=> {
      // browserWindow.close()
      coscript.setShouldKeepAround(false); // allow session tear down
      throw(null) // force tear down start
    }
    setTimeout(tearDown, 5000) // alt in menu/keyboard/window close trigger
    ```

- install script by link
  - `npm run build`
  - `ln -s "$(pwd)/$plugin.sketchplugin" ~/Library/Application\ Support/com.bohemiancoding.sketch3/Plugins/$plugin.sketchplugin`
  - start sketch *// restart not necessary if already started*
- cycle
  - edit source (eg. change `ver` in demo above)
  - `npm run build`
  - start script from sketch (through shortcut or sketch runner, eg `cmd-', enter` after first time)
    *// if getting previous ver, it hadn't finished tear down, or AlwaysReloadScript flag not set*
  - trigger tearDown *// eg. await timeout, before running again*

---

Would be neat to have the `npm run start` to automatically trigger these steps instead of doing whatever it does now (that, for me, either crashes, serves the old plugin, or takes forever to build (?) after some time).
- plugin source code change detected (not webview, that's for webpack to take care of)
- send some command either to sketch or directly to the plugin to trigger the plugin's tearDown
- re-build
- trigger the plugin start comment in sketch (if provided like `npm run start --run` or `npm run start --run some-action-defined-in-manifest`)

**Comment by mathieudutour** [comment]
Fixed in 0.7.0 :tada: (but it will require Sketch 56)
